loadable modules: add support for loading modules on Xtensa

lyakh opened the pull request:
This is based on an earlier version of #57086 . The latter has a few more commits now, but the final state isn't very different from my snapshot - @teburd has fixed a couple of bugs that I also found and fixed in this PR, added support for some ARM specific features, and started work on Xtensa relocation code, which isn't used here yet. So far this code can load shared objects, built as PIC, which don't need relocation. This is still WiP, I expect comments, fixes, and a couple more iterations of this.

lyakh committed:
llext: export some symbols

Export some symbols for loadable modules. Also add an
EXPORT_SYSCALL() helper macro for exporting system calls by their
official names.

Signed-off-by: Guennadi Liakhovetski <guennadi.liakhovetski@linux.intel.com>

lyakh committed:
llext: add support for shared objects

Add support for linking PIC shared object, which only require
linking, using their PLT and GOT lists and don't need any relocation
otherwise.

Signed-off-by: Guennadi Liakhovetski <guennadi.liakhovetski@linux.intel.com>

teburd commented:
I've worked on incorporating the EXPORT_SYMBOL work from this PR, which was great. I've reused some existing macros/linker scripts that are useful beyond just ace to do that. Iterable sections and common-ram.ld for now.

Ideally in the base image the symbol table would be const and placed in rom for most systems, and in loadable modules perhaps a custom linker section could be constructed and then imported to determine which symbols are actually intentionally exported as opposed to the ones marked extern as I've done so far.

For loadable modules we likely want the symbol table to be still placed in RAM so that it can be updated on relocation.

lyakh commented:
> I don't think generally we can rely on module_peek though.

@teburd I thought it would be a bit controversial. But so far I see only one implementation of module operations in your PR - one for buffers. And in it your `module_buf_read()` does just a `memcpy()`, so it does assume direct access to that memory. And as you see `.module_peek()` is extremely convenient - it eliminates the need to copy large memory areas of unknown size, both saving memory and time. I'd very much like to find a way to cleanly integrate it.

teburd commented:
> > I don't think generally we can rely on module_peek though.
> 
> @teburd I thought it would be a bit controversial. But so far I see only one implementation of module operations in your PR - one for buffers. And in it your `module_buf_read()` does just a `memcpy()`, so it does assume direct access to that memory. And as you see `.module_peek()` is extremely convenient - it eliminates the need to copy large memory areas of unknown size, both saving memory and time. I'd very much like to find a way to cleanly integrate it.

So just to clarify my thinking here...

Not all streams will come from memory that can be mapped into an addressable space, its very likely some elfs will come from off-chip flash that can be read from and copied out but not mapped.

At the same time I completely agree that if you have memory that can be directly mapped, and you happen to have an elf file that is padded and aligned appropriately... as is I believe your case, it should be entirely supported to directly map the elf.

In the buffer case, its pretty lame that there's allocations for things like elf headers and section headers when its... right there and can be pointed at.

So to support all of these cases I think a little bit of rethinking on the module stream interface is needed to support it.

Basic idea would be to modify module_stream_read to take... the stream, memory kind (rx, ro, rw, metadata), and length to read as parameters, while returning a void *.

e.g.

``` c
void* module_read(struct module_stream *ms, enum module_mem_kind *mk, size_t sz);
```

The stream itself can be the allocator then when it needs to be, or it can simply pass back a pointer into the elf. And since the stream API is overrideable it means any of the above scenarios is easily dealt with, leaving the load and link functions to do what the need to without worrying about it or having branching all over the place.

This is likely worth doing *after* the initial PR though as the initial PR is already quite close and mostly needs one other architecture to work imo.

lyakh committed:
llext: add a function for finding ELF sections

Applications can use custom ELF sections for their own purposes, add
a function for finding them.

Signed-off-by: Guennadi Liakhovetski <guennadi.liakhovetski@linux.intel.com>

lyakh committed:
llext: make local relocations optional

Some applications can decide to link their loadable objects for
exactly the same addresses, where they will be loaded. In those cases
local relocations aren't needed any more and can in fact break the
object if applied while the object is in a temporary storage. Add a
parameter to skip such local relocations.

Signed-off-by: Guennadi Liakhovetski <guennadi.liakhovetski@linux.intel.com>

lyakh commented:
> ```
> void* module_read(struct module_stream *ms, enum module_mem_kind *mk, size_t sz);
> ```

`mk` doesn't have to be a pointer, does it? Otherwise - yes! One *could* also simplify the API a bit more by removing `module_seek()` and adding an offset to `module_read()` but that's optional - I'm fine either way.

> The stream itself can be the allocator then when it needs to be, or it can simply pass back a pointer into the elf. And since the stream API is overrideable it means any of the above scenarios is easily dealt with, leaving the load and link functions to do what the need to without worrying about it or having branching all over the place.
> This is likely worth doing after the initial PR though as the initial PR is already quite close and mostly needs one other architecture to work imo.

As also just discussed offline, internal allocation is difficult because it's unknown when to free that memory. I think we need an additional memory pointer for the call. The caller will provide a buffer - be it allocated, on stack or static. And if the caller knows, that they can peek at the image, they'll supply `NULL` for the buffer

Otherwise this version now works both with twister with the added here Xtensa ELF object, and with SOF.

lyakh committed:
llext: xtensa: add support for local symbol relocations

Add support for relocating local symbols, as specified in the
.rela.dyn section.

Signed-off-by: Guennadi Liakhovetski <guennadi.liakhovetski@linux.intel.com>

lyakh committed:
llext: make buffer access functions accessible externally

llext_seek(), llext_read() and llext_peek() are needed outside of the
extension code too, move them to a header.

Signed-off-by: Guennadi Liakhovetski <guennadi.liakhovetski@linux.intel.com>

lyakh committed:
llext: add Xtensa test support

Add support for running a modular "Hello world" example on Xtensa.

Signed-off-by: Guennadi Liakhovetski <guennadi.liakhovetski@linux.intel.com>

lyakh committed:
llext: add support for global functions

Loadable modules can contain global (not "static") functions, even if
they aren't exported for use by other modules, e.g. when a module is
built from multiple .c files. Such functions are then also included
in link tables and have to be re-linked.

Signed-off-by: Guennadi Liakhovetski <guennadi.liakhovetski@linux.intel.com>

lyakh commented:
This now temporarily includes #64381 until it's merged

lyakh commented:
This is an interesting failure "Illegal use of the EPSR" https://github.com/zephyrproject-rtos/zephyr/actions/runs/6665109962/job/18114283475?pr=62433 @teburd any ideas?

lyakh committed:
llext: add reference counting

Extend the llext_load() / llext_unload() API to let it be called
repeatedly for the same extension to increment or decrement its
reference counter respectively. We use a mutex to protect the counter
and make both llext_load() and llext_unload() return the use-count to
let the caller identify when the first loading and the last unloading
took place.

Signed-off-by: Guennadi Liakhovetski <guennadi.liakhovetski@linux.intel.com>

teburd commented:
> This is an interesting failure "Illegal use of the EPSR" https://github.com/zephyrproject-rtos/zephyr/actions/runs/6665109962/job/18114283475?pr=62433 @teburd any ideas?

No ideas from me no, I can try and take a look at this closer today. There's another issue on cortex-m7 as well with dcache involved.

It might be worth trying to break this PR up a bit into some small PRs that get merged incrementally. PLT/GOT support with -fPIC, xtensa support, and maybe lastly the peek + copy-free usage perhaps?

lyakh commented:
> It might be worth trying to break this PR up a bit into some small PRs that get merged incrementally. PLT/GOT support with -fPIC, xtensa support, and maybe lastly the peek + copy-free usage perhaps?

@teburd can we do it in a different order - first add `llext_peek()` and then add PLT? I'd prefer to add Xtensa support on top of a peek-enabled implementation directly?

lyakh committed:
llext: fix a memory leak in an error case

If a function fails it should release all the resources it has
managed to acquire. Fix llext_load() to free memory that it has
allocated in case of an error.

Signed-off-by: Guennadi Liakhovetski <guennadi.liakhovetski@linux.intel.com>

lyakh committed:
llext: add support for exporting symbols from extensions

Extensions should be able to selectively export their global symbols.
Add a LL_EXTENSION_SYMBOL() macro for that. Change the present
.sym_tab to be a temporary symbol table of all global symbols in an
extensions, used only during linking for internal purposes. Add a new
.exp_tab symbol table to store symbols, exported by an extension
permanently.

Signed-off-by: Guennadi Liakhovetski <guennadi.liakhovetski@linux.intel.com>

lyakh committed:
llext: clarify section map allocation size

Use an element size explicitly when calculating the array size and
use the calculated size for memset().

Signed-off-by: Guennadi Liakhovetski <guennadi.liakhovetski@linux.intel.com>

teburd commented on include/zephyr/llext/llext.h:
```diff
@@ -88,19 +101,22 @@ struct llext *llext_by_name(const char *name);
  * @retval -EINVAL Invalid ELF stream
  */
-int llext_load(struct llext_loader *loader, const char *name, struct llext **ext);
+int llext_load(struct llext_loader *loader, const char *name, struct llext **ext, bool do_local);
```
I feel like we're likely to have more various options/flags/etc for configuring how loading and linking occurs, can we move the do_local bool the loader struct itself?

pillo79 replied:
The `do_local` argument seems to be only used to prevent `llext_link_plt` from calling `arch_elf_relocate_local`, which is useful for Xtensa but has a weak stub for other arches. Is this argument really needed?

lyakh replied:
@teburd I'm trying to limit the scope of variables. This parameter is only needed within this single loading call. We now store the llext context permanently - from loading until unloading, but this parameter is only needed during loading. If we get more of such short-lived parameters, I'd maybe add a temporary loader-context structure, that isn't stored permanently and thus is just an object on stack in `llext_load()`

teburd commented on lib/libc/minimal/source/string/string.c:
I'd note this only exports the symbols for the minimal lib C, it might make more sense to create a libc_export.c somewhere with these exports such that any libc (minimal, newlib, picolib) the symbols are able to be found.

cfriedt replied:
Given that the C library symbols are global and not really specific to any libc, couldn't they just be exported from anywhere? E.g. the core llext code?

I don't really see a reason to tie in any llext code here at all.

teburd commented on subsys/llext/llext.c:
```diff
@@ -709,48 +909,67 @@ int llext_load(struct llext_loader *ldr, const char *name, struct llext **ext)
 		(*ext)->name[sizeof((*ext)->name) - 1] = '\0';
+		(*ext)->use_count++;
+
 		sys_slist_append(&_llext_list, &(*ext)->_llext_list);
```
Since we have a mutex protecting the removal we can reuse it presumably to protect appending a module here

lyakh replied:
the purpose of locks has to be strictly defined. So far `llext_lock` is used to protect reference counting. If we also want to use it to protect the list of modules, it has to be carefully considered and - if accepted - properly documented. I haven't done that yet, therefore it's still a "TODO" / "FIXME." In fact so far we don't use that list of modules, but we probably will need it when modules begin using symbols, exported by other modules.

teburd commented on subsys/llext/llext.c:
```diff
@@ -709,48 +909,67 @@ int llext_load(struct llext_loader *ldr, const char *name, struct llext **ext)
+		free(tmp, ctx);
+	}
+
+	/* FIXME: protect the global list */
```
Looks like its in a mutex now, so is protected

lyakh replied:
only here, but it's still unprotected in most locations.

lyakh replied:
@teburd please check the update, I added protection to the list too

teburd commented on subsys/llext/llext.c:
```diff
@@ -709,48 +909,67 @@ int llext_load(struct llext_loader *ldr, const char *name, struct llext **ext)
+	}
+
+	if (free) {
+		free(tmp, ctx);
```
Could all the params given to unload for freeing be given to the loader and saved as part of the module instead? It's unclear otherwise how the free fn and ctx are to be used here to me.

pillo79 replied:
Agree with the above - even more, the use case for this feels unclear to me:
* Before calling `llext_unload` on an used llext, I expect the caller to have nothing more to do with the llext, so for a single user it feels redundant;
* Since now we have ref counting, if some llext is shared by 2 users, and each needs its own callback, this should be saved as a list so every user's callbacks get called. 

Still, I don't see a clear use for this - I keep thinking about calling special "init" and "exit" functions _in the llext_ (Linux module style)...

lyakh replied:
@teburd I had that in my first version, but then I realised that those parameters are only used once here for freeing, and we do call `llext_unload()` explicitly anyway, so there's no point in storing those values. We can just pass them directly here.

lyakh replied:
@pillo79 let me explain a bit. With SOF we have an auxiliary extension management layer. Its tasks include (1) loading extensions from an external source into slow but abundant memory, (2) copying them for use into fast but scarce memory, (3) freeing back that memory when no longer in use.
The flow is then as follows:
1. load from source during startup
2. when needed:
2.1. call `llext_load()` to parse and link ELF
2.2. check return code to see if it's the first use
2.3. if so, copy to fast RAM
3. when unneeded:
3.1. call `llext_unload()`
3.2. if that was the last use, our callback is called, which frees the fast memory

I think in step 3.2 above we could do the same as in 2.2 - check return code and free memory if last. Would that be a better solution? This also explains, that we don't have different callbacks for multiple users, but only one of them. It's also correct, that currently we don't use the `llext` context in our callback, but I thought that in the future that could change so I implemented it this way. But if we switch to return value checking, then, of course, that context will be gone anyway after `llext_unload()` returns.

lyakh replied:
@pillo79 ok, I'll try to switch over to a return code then, stay tuned :-)

teburd commented on tests/subsys/llext/hello_world/hello_world.c:
```diff
@@ -12,13 +12,15 @@
 	printk("hello world\n");
 	printk("A number is %lu\n", number);
 }
+LL_EXTENSION_SYMBOL(hello_world);
```
Great!

pillo79 replied:
Good idea, but this makes building an llext more complicated. Not impossible of course, but I feel some comments should be added to the documentation.

teburd reviewed (commented):
Looks really nice, a few structural things I believe could keep the API the same while providing the added features.

do_local especially I believe should go in the loader struct, its very likely more flags will be coming to decide e.g. what sections to copy or not, even if peek is supported.

pillo79 commented on subsys/llext/shell.c:
There's a call to `llext_unload()` in this file that needs the new arguments (or not, if they are moved somewhere else :wink:).

cfriedt reviewed (dismissed):
Do we really need to add llext code to the implementations of common libc string functions?

Given that the symbols are global and not specific to any libc, I would imagine they could be exported anywhere - e.g. by some llext common code.

lyakh committed:
llext: remove llext_list()

llext_list() is an exported function that returns a pointer to the
llext internal extension list. That list should only be accessible
directly inside llext, while holding a lock. Remove the function.

Signed-off-by: Guennadi Liakhovetski <guennadi.liakhovetski@linux.intel.com>

lyakh committed:
llext: remove redundant initialisation

a new llext object is completely initialised with zeros after
allocation, no need to additionally set members of an embedded into
it array to NULL.

Signed-off-by: Guennadi Liakhovetski <guennadi.liakhovetski@linux.intel.com>

lyakh committed:
llext: protect the global llext list

Use an existing mutex to also protect the global llext list.

Signed-off-by: Guennadi Liakhovetski <guennadi.liakhovetski@linux.intel.com>

lyakh commented:
@cfriedt @teburd @pillo79 I've fixed libc exports and changed `llext_unload()` to use the return code. Please check, if you agree with my explanation for other comments or if anything else should be changed

pillo79 commented:
Looking good to me! Last nitpick... can you explain when [the `do_local` argument](https://github.com/zephyrproject-rtos/zephyr/pull/62433#discussion_r1397702923) is strictly necessary? It looks to me as a very specific detail that may not be useful to everyone in the generic llext API. 
 * If it can be wrapped inside `llext_load` as you were saying, I'm all for it!
 * Another option is, can we switch it to a generic `flags` variable and define this as (the first) target-specific flag?

cfriedt commented on kernel/thread.c:
```diff
@@ -141,6 +142,7 @@ bool k_is_in_isr(void)
 {
 	return arch_is_in_isr();
 }
+EXPORT_SYMBOL(k_is_in_isr);
```
Again, this is a global symbol - there is no implicit dependency on llext for `k_is_in_isr()`.

I can understand the case where a particular private or static function needs to be exported for llext (for visibility), but for globals, standard C functions, can we just put them into `llext_export.c` as well?

lyakh replied:
well, I think I'd disagree here. Think about adding a new function that you want to export. You add it where it belongs and you just add that one single line to that file to indicate that it's exported. You don't want to go and edit a common global file for all these exports. Also this is the same as what the Linux kernel does. We agreed with libc symbols because those functions can have multiple implementations, but "normal" functions, that only have one implementation each, should export the symbol themselves.

cfriedt replied:
Linux is built around the premise of *most* optional components being tristate selectable. Zephyr is not. There is a huge difference.

What you're describing is a treewide change introducing a dependency on a header that is experimental. There is a process for that. Personally, I would prefer to see a less invasive approach for this feature until it is stable.

It would be different if it were just part of the regular Zephyr API and available via `<zephyr/kernel.h>` or something along those lines, but realistically, it is still experimental.

cfriedt commented on lib/os/assert.c:
```diff
@@ -53,3 +54,4 @@ void assert_print(const char *fmt, ...)
 
 	va_end(ap);
 }
+EXPORT_SYMBOL(assert_print);
```
Again, these are global symbols - there is no implicit dependency on llext for assert functionality.

I can understand the case where a particular private or static function needs to be exported for llext (for visibility), but for globals, standard C functions, can we just put them into `llext_export.c` as well?

cfriedt commented on subsys/logging/log_msg.c:
```diff
@@ -270,6 +271,7 @@ void z_impl_z_log_msg_static_create(const void *source,
 
 	z_log_msg_finalize(msg, source, out_desc, data);
 }
+EXPORT_SYMBOL(z_impl_z_log_msg_static_create);
```
Again, this is a global symbol - there is no implicit dependency on llext for logging.

Can we put this into `llext_export.c` as well?

lyakh committed:
llext: remove llext list scanning from shell.c

The llext list should be internal to llext.c, remove its scanning
from shell.c, export a function for that instead.

Signed-off-by: Guennadi Liakhovetski <guennadi.liakhovetski@linux.intel.com>

teburd commented:
Looking pretty good, I do wonder about the mutex being used inside of the llext_by_name/llext_by_index, this could lead to some confusing bugs if used with a shell or external management utility through like ble or something.

Maybe exposing the lock through a a public llext_lock() llext_unlock() pair of functions so that when using the list or iterating through it, the lock can be taken and you are guaranteed the list isn't mutated underneath you.

E.g. imagine someone is iterating through the extensions (llext_by_index) while another core/thread/context is simultaneously updating it.. some very odd behavior would occur.

Vs in the shell when listing things a lock can be taken/released around the usage of llext and the entire problem avoided.

pillo79 commented:
> Maybe exposing the lock through a a public llext_lock() llext_unlock() pair of functions so that when using the list or iterating through it, the lock can be taken and you are guaranteed the list isn't mutated underneath you.

I personally don't like this, it would expose an internal implementation detail only to allow `llext_by_index` to be used to walk the list safely.

> imagine someone is iterating through the extensions (llext_by_index) while another core/thread/context is simultaneously updating it.. 

I'd go a step further - I appreciate `llext_by_name` as you can get a reference to a specific llext you hopefully know it's "locked" in memory, but `llext_by_index` looks like a source of problems as the indexes may change all the time. 
Would it be possible to provide, e.g., an `llext_list_walk()` API with a callback instead? Then locking could be performed internally in a safe way...

lyakh commented:
> Looking good to me! Last nitpick... can you explain when [the `do_local` argument](https://github.com/zephyrproject-rtos/zephyr/pull/62433#discussion_r1397702923) is strictly necessary? It looks to me as a very specific detail that may not be useful to everyone in the generic llext API.
> 
>     * If it can be wrapped inside `llext_load` as you were saying, I'm all for it!
> 
>     * Another option is, can we replace that with a generic `flags` argument, and define this as the first of hopefully many flags?

Ok, as explained, I'd rather not add that parameter to `struct llext`, but I'd be ok with making it a flag bitmask. Although - basically for me this is a "non-persistent" parameter as opposed to `struct llext *ext` which is persistent. If we get any more of such temporary parameters we can make something like `struct llext_load_param`. I thought as long as we only know of one such parameter we could pass it explicitly, but if preferred, I can make such a structure with a single member already now.

pillo79 commented:
I was asking about the technical reason for requiring the argument in the first place :slightly_smiling_face:, but really I'm just trying to avoid changing the public API too often.
I'm not convinced `do_local` should be a first-class argument as the other two; if needed, I'm fine with adding either a bitmask or a separate struct, as long as zero (or a zero-initialized struct) is a sane default. (Should it be a `skip_local` perhaps?)

lyakh commented:
> I was asking about the technical reason for requiring the argument in the first place 🙂

@pillo79 ah, sorry. The reason is that as I mentioned before SOF stores extensions permanently in slow abundant memory. When those extensions are first used we run `llext_load()` on that memory, performing all the linking and relocating. Then if we destroy audio pipelines using that extension, its use-count goes to zero, we unload it and "forget about it," but the relocations that we've done on the module - all the resolved internal references - remain resolved, so there's no need to resolve them again if we next time need to use that extension again. Moreover, since some of those relocations are calculated as offsets, adding offsets to already correct addresses breaks them. So we skip performing local relocations on subsequent loads.
Ok, let's use a structure for loader private parameters too, will update tomorrow.

pillo79 commented:
Thanks for the explanation, clear as always!

> So we skip performing local relocations on subsequent loads.

Sounds to me you really want to store this as a `reloc_done` flag in the loader struct :wink: and handle it automatically in the llext code: it'd be helpful to all `buf_loader` users where loading may have patched the source RAM buffers.

lyakh commented:
> Sounds to me you really want to store this as a `reloc_done` flag in the loader struct 😉 and handle it automatically in the llext code: it'd be helpful to all `buf_loader` users where loading may have patched the source RAM buffers.

@pillo79 the llext context is destroyed between those two calls. The core doesn't know that it's the same buffer and that it has already relocated it.

lyakh committed:
llext: fix read-only extension image

When using the LLEXT buffer loader we now avoid copying extensions
from storage to allocated memory by pointing directly into the stored
image. We then also perform linking and relocation in that memory,
which modifies its contents. However, this is impossible if that
storage is read-only. Add a Kconfig flag to distinguish between
writable and read-only storage types. Also use that flag to decide,
whether the extension image in test_llext_simple.c should be defined
as const or not.

Signed-off-by: Guennadi Liakhovetski <guennadi.liakhovetski@linux.intel.com>

pillo79 commented on include/zephyr/llext/llext.h:
```diff
@@ -78,6 +86,30 @@ sys_slist_t *llext_list(void);
+ * These are parameters, not saved in the permanent llext context, needed only
+ * for the loader
+ */
+struct llext_load_param {
+	/** Should local relocation be performed */
+	bool relocate_local;
+};
```
Most users will want to set this, so a non-zero value should be used as a default. In this case I suggest adding a `LLEXT_LOAD_PARAM_DEFAULT` macro so it's easy to initialize the struct correctly, and it's future proof when other special flags are added.

lyakh replied:
@pillo79 how about I allow `NULL` and assume `true` by default? I do have to check for `NULL` anyway, it's missing now

pillo79 replied:
Nice addition! But I still think it would be helpful to have a future-proof public default (currently as small as `{ true }` :slightly_smiling_face: ), somewhat like the LLEXT_BUF_READER macro. 
If I were using NULL and later I wanted to customize some aspect of it, I could see defaults clearly documented. Otherwise I'd have to go and find out from the source what "NULL" actually meant.

cfriedt commented:
I don't want to block unnecessarily, so sorry for the NAK. It would be great to have @dcpleung or @andyross comment here too.

* `EXPORT_SYMBOL(k_is_in_isr)` looks fine
* `EXPORT_SYMBOL(z_impl_k_mutex_unlock)` leaves me feeling a bit uneasy
* `EXPORT_SYMBOL(z_impl_z_log_msg_static_create)`, similarly

Was there already a discussion / resolution at the arch meeting about it? Sorry if I missed that.

For `__syscall` functions, with static inline / pinned function implementations, are we just trying to optimize-out the pinned function here?

I can sort of understand that, but I feel like it would almost be nicer to abstract things a bit more in the latter case by not explicitly naming the `z_impl_` part (which is implementation detail).

```cpp
#define EXPORT_SYSCALL(_call) EXPORT_SYMBOL(z_impl_ ## _call)
...
EXPORT_SYSCALL(k_mutex_unlock);
```

This would at least gave the impression that we were only dealing with public-facing API, which is **extremely** important for modules / llext. In Linux, for example, there is the mantra of never breaking user-facing APIs. So user-facing APIs have some permanence. Of course, we deprecate some symbols and have a process for that..

Again, sorry if this was already discussed.

lyakh commented:
Update 20231128:
- added `EXPORT_SYSCALL()` (thanks @cfriedt)
- added a default `struct llext_load_param` initialiser (thanks @pillo79)
- made read-only llext storage a Kconfig option (thanks @teburd)

lyakh commented:
@cfriedt any more blockers?

cfriedt commented:
None on my end, but there may be a couple of still relevant compliance issues. Likely the EXPORT_SYMBOL() warnings can be ignored.

teburd reviewed (approved):
We might need a few CI checks fixed, let me look, but approved

nashif commented:
overriding checkpatch as false negatives.
